rust/lance-encoding/src/encodings/logical/primitive.rs

```diff
@@ -563,7 +584,374 @@ impl FieldEncoder for PrimitiveFieldEncoder {
+        let mut data_buffer = Vec::with_capacity(
+            miniblocks.data.len()
+                + bytes_rep
+                + bytes_def
+                + max_bytes_repdef_len
+                + MINIBLOCK_MAX_PADDING as usize,
+        );
+        let mut meta_buffer = Vec::with_capacity(miniblocks.data.len() * 2);
+
+        let mut value_offset = 0;
+        for ((chunk, rep), def) in miniblocks.chunks.into_iter().zip(rep).zip(def) {
+            let chunk_bytes = chunk.num_bytes as u64 + rep.len() as u64 + def.len() as u64 + 4;
+            assert!(chunk_bytes <= 16 * 1024);
+            assert!(chunk_bytes > 0);
+            // We subtract 1 here from chunk_bytes because we want to be able to express
+            // a size of 24KiB and not (24Ki - 6)B which is what we'd get otherwise with
+            // 0xFFF
+            let divided_bytes = (chunk_bytes - 1).div_ceil(MINIBLOCK_SIZE_MULTIPLIER);
+            let pad_bytes = (MINIBLOCK_SIZE_MULTIPLIER * divided_bytes) - (chunk_bytes - 1);
+
+            let metadata = ((divided_bytes << 4) | chunk.log_num_values as u64) as u16;
+            meta_buffer.extend_from_slice(&metadata.to_le_bytes());
+
+            assert!(rep.len() < u16::MAX as usize);
+            assert!(def.len() < u16::MAX as usize);
+            let bytes_rep = rep.len() as u16;
+            let bytes_def = def.len() as u16;
+
+            data_buffer.extend_from_slice(&bytes_rep.to_le_bytes());
+            data_buffer.extend_from_slice(&bytes_def.to_le_bytes());
+
+            data_buffer.extend_from_slice(&rep);
+            data_buffer.extend_from_slice(&def);
+
+            let num_value_bytes = chunk.num_bytes as usize;
+            let values =
+                &miniblocks.data[value_offset as usize..value_offset as usize + num_value_bytes];
+            data_buffer.extend_from_slice(values);
+
+            data_buffer.extend(iter::repeat(0).take(pad_bytes as usize));
+
+            value_offset += num_value_bytes as u64;
+        }
+
+        (
+            LanceBuffer::Owned(data_buffer),
+            LanceBuffer::Owned(meta_buffer),
+        )
+    }
+
+    /// Compresses a buffer of levels
+    ///
+    /// TODO: Use bit-packing here
+    fn compress_levels(
+        levels: Option<LevelBuffer>,
+        num_values: u64,
+        compression_strategy: &dyn CompressionStrategy,
+        chunks: &[MiniBlockChunk],
+    ) -> Result<(Vec<LanceBuffer>, pb::ArrayEncoding)> {
+        if let Some(levels) = levels {
+            debug_assert_eq!(num_values as usize, levels.len());
+            // Make the levels into a FixedWidth data block
+            let mut levels_buf = LanceBuffer::reinterpret_vec(levels);
+            let levels_block = DataBlock::FixedWidth(FixedWidthDataBlock {
+                data: levels_buf.borrow_and_clone(),
+                bits_per_value: 16,
+                num_values,
+            });
+            let levels_field = Field::new_arrow("", DataType::UInt16, false)?;
```

Comment:
Is it worth pulling this out into a static?

Reply:
I'll do that in the read PR